src/ControlSystem/ControllerPND.cpp

```diff
@@ -0,0 +1,89 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ControlSystem/ControllerPND.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+
+namespace Controllers {
+
+template <size_t deriv_order>
+DataVector PND<deriv_order>::operator()(
+    const DataVector& timescales,
+    const std::array<DataVector, deriv_order + 1>& q_and_derivs,
+    const double q_dt, const double deriv_dt) const noexcept {
+  // helper lambda for computing the binomial coefficients
+  auto binomial = [](size_t N, size_t k) {
+    return falling_factorial(N, k) / factorial(k);
+  };
+
+  // The coefficients a0_{k} are chosen such that the system is critically
+  // damped, which are given by: a0_{k} = (N choose k) timescales^{k-N}
+  std::array<DataVector, deriv_order> coefs0;
+  const DataVector inv_timescales = 1.0 / timescales;
+  DataVector tau = inv_timescales;
+  for (size_t i = deriv_order; i-- > 0;) {
+    gsl::at(coefs0, i) = binomial(deriv_order, i) * tau;
+    // add another power of timescales
+    tau = tau * inv_timescales;
+  }
+
+  // If an averager is used for q_and_derivs, there is an induced time offset,
+  // `t_offset`, due to the time-weighted averaging. Therefore, the q_and_derivs
```

Comment:
Is there a derivation/explanation/reference to a paper that explains why there is an induced time offset? I don't know whether it's a hard thing to show or if it's self evident to someone to who uses averagers often

Reply:
After thinking about it more I understand now that the offset is a result of average over qs and derivs of qs of previous timesteps, so this averaged value corresponds to a quantity that lags behind the current time. Can I suggest a wording like "Obtaining a q_and_derivs  by averaging over previous timesteps results in a q_and_derivs that lags behind the current values q and its derivatives." ?